### PR TITLE
Introduce Rust symbol mangling scheme.

### PR DESCRIPTION
**Pull request by eddyb**
This is an implementation of a "feature-complete" Rust mangling scheme, in the vein of rust-lang/rfcs#2603 ~~- but with some differences, see https://github.com/rust-lang/rfcs/pull/2603#issuecomment-458410463 for details~~ (@michaelwoerister integrated my proposed changes into the RFC itself).

On nightly, you can now control the mangling scheme with `-Z symbol-mangling-version`, which can be:
* `legacy`: the older mangling version, still the default currently
* `v0`: the new RFC mangling version, as implemented by this PR

To test the new mangling, set `RUSTFLAGS=-Zsymbol-mangling-version=v0` (or change [`rustflags` in `.cargo/config.toml`](https://doc.rust-lang.org/cargo/reference/config.html#configuration-keys)). Please note that only symbols from crates built with that flag will use the new mangling, and that tool support (e.g. debuggers) will be limited initially, and it may take a while for everything to be upstreamed. However, `RUST_BACKTRACE` should work out of the box with either mangling version.

<hr/>

The demangling implementation PR is https://github.com/alexcrichton/rustc-demangle/pull/23
~~(this PR already uses it via a git dependency, to allow testing)~~.

Discussion of the *design* of the mangling scheme should still happen on the RFC, but this PR's specific implementation details can be reviewed in parallel.

*Notes for reviewers*:
* ~~only the last 6 commits are specific to this branch, if necessary I can open a separate PR for everything else (it was meant to be its own small refactoring, but it got a bit out of hand)~~
~~based on #58140~~
* the "harness" commit is only there because it does some extra validation (comparing the demangling from `rustc-demangle` to the compiler's pretty-printing, adjusted slightly to produce the same output), that I would like to try on crater
* ~~there is the question of whether we should turn on the new mangling now, wait for tools to support it (I'm working on that), and/or have it under a `-Z` flag for now~~ (we're gating this on `-Z symbol-mangling-version=v0`, see above)

r? @nikomatsakis / @michaelwoerister cc @rust-lang/compiler

### CODE REVIEW

**Comment by eddyb** [comment]
So that's probably because the checking is quite thorough (therefore a bit slow) - I don't want to compromise on that, is there any way to temporarily raise the limit?

cc @rust-lang/infra

**Comment by eddyb** [comment]
@michaelwoerister @alexcrichton The `run-make-fulldeps/a-b-a-linker-guard` failure is because we used to hash the signature in a symbol name, but now we don't have any hashing, and the signature is not explicitly encoded, so the symbol is the same. We can:
* remove the test, or adjust it so the symbols differ in some other way)
* add the fn signature, or a hash of it, to the symbol

If we add the actual input/output types, it will significantly inflate symbol name sizes, whereas a hash would be a constant overhead (but @michaelwoerister's RFC tries to avoid hashes for anything other than crate roots).

IMO ideally the entire dylib would be versioned in some way, so a new build would be incompatible - IIRC we have something (SVH?) for detecting this when `rustc` is invoked, can it be placed in the dylib in a way the loader would check it, and error on mismatch?

**Comment by pietroalbini** [comment]
Do you mean only for this try build or for a while? We'd have to ask Travis Support either way.

**Comment by eddyb** [comment]
@pietroalbini Only for this build, mostly for running crater. I suppose I can disable some of the work not needed to check symbols, but the roundtrip check is probably the most expensive thing anyway.

I guess I'll see how long it takes for the PR build to fail again, with as much as possible disabled.

**Comment by eddyb** [comment]
> I guess I'll see how long it takes for the PR build to fail again, with as much as possible disabled.

Time-to-fail is not as useful as I hoped, so I turned off the failing test (https://github.com/rust-lang/rust/pull/57967#issuecomment-458626105).

@bors try

**Comment by eddyb** [comment]
@craterbot run start=master#c1c3c4e95b69dfeaca5c5db6c622d7f90ad30a54 end=try#7fea0229842e79db1159ea58ad9a653b288efee9

**Comment by michaelwoerister** [comment]
Thanks, @eddyb! That's an epic amount of refactorings `:)` I think the new mangling scheme should be opt-in via a `-Z` flag since the RFC hasn't been accepted yet.

It might make sense to split the `ppaux` refactorings into a separate PR so things are less likely to bitrot.

**Comment by michaelwoerister** [comment]
I don't know yet what to do about the failing test. I believe that the same thing would be caught if things went through cargo, because then the two crates would have a different disambiguator. Or at least they probably *should* have a different disambiguator.

**Comment by alexcrichton** [comment]
According to [this link](https://crater-reports.s3.amazonaws.com/pr-57967/try%237fea0229842e79db1159ea58ad9a653b288efee9/gh/jstzwj.gitcube/log.txt)

```
[INFO] [stderr] error: symbol `_RINhNyCsfOXHOjakNhA_4core3mem7size_ofINcNsB4_6option6OptionINyNyCsb07I8x9r9XL_9actix_web9extractor4FormINxNxNyNeCs93O65jcFPR3_3std11collections4hash3map7HashMapNaNqCskZeSixylFoE_5alloc6string6StringB2y_EEEECs9uonOkHWQaN_7gitcube` is already defined
```

that's quite the mouthful!

**Comment by eddyb** [comment]
@alexcrichton @michaelwoerister I think we should pass such symbols through `rustc-demangle` - we already demangle symbols for `--emit=llvm-ir`, to generate helpful comments on defintions and call/invoke instructions.

**Comment by eddyb** [comment]
@rust-timer build 7fea0229842e79db1159ea58ad9a653b288efee9

**Comment by eddyb** [comment]
Minimal reproduction:
```rust
fn size_of_val<T>(_: &T) -> usize {
    std::mem::size_of::<T>()
}

struct Foo(i64);

pub fn size_of_foo() -> usize {
    size_of_val(&Foo(0))
}
pub fn size_of_foo_ctor() -> usize {
    size_of_val(&Foo)
}
```

Surprised we don't have a test for this! So the problem is that I'm stripping a `StructCtor` component from the symbol without modifying the rest of the symbol in any way.

I think we should just encode it in some way (the same way closures are?) and have the demangler ignore it (if we even want it to be ignored).

**Comment by eddyb** [comment]
I pushed a fix for the `actix-web` / `rusturn` failures.

`css-modules` [appears to be inspecting mangled symbols](https://github.com/measlytwerp/css-modules-rs/blob/fdae5d9188efe668a3bb767698baa38ca0f012df/src/lib.rs#L49) and should be fine once the `backtrace` crate updates to a future `rustc-demangler` version that includes the new mangling.

**Comment by michaelwoerister** [comment]
I advise against calling something the "new" or "modern" xyz in the code. Something might change in the future and the "new" thing is really the legacy thing but it's still called "new_xyz" everywhere. "v2" or "v2019" or some other absolute name is more future proof.

**Comment by eddyb** [comment]
@michaelwoerister I want to rely on the optional version, and maybe just call the old one `legacy`, whereas the new one is `v0`.
(The `new` module is only in the temporary setup, which I need to redo to get this into a mergeable shape)

Or do we want to retroactively call the old one `v0` and the new one `v1`, which would mean the next mangling format will start with `_R2`?

I think it would be *far* more confusing to call it something including a number, and have a number in the symbols, but not have those match.

There's also `modern` in the demangler, should probably use the same choice of names there too.

**Comment by michaelwoerister** [comment]
I'd be fine with `legacy` for the current mangling and `v0` for the new one.

**Comment by Dylan-DPC** [comment]
ping from triage @eddyb you have conflicts to resolve

**Comment by michaelwoerister** [comment]
I'm wondering what the way forward is here. I think we can land it under a `-Z` flag. Is there value in landing the "temporary" commits?

**Comment by eddyb** [comment]
@michaelwoerister I'm preparing a version with the flag, now that #59343 is approved.

I'll revert most of the changes not useful to the `-Z symbol-mangling=...` flag, but keep them in the commit history in case we need to reuse some of the testing harness.

**Comment by Dylan-DPC** [comment]
ping from triage @eddyb you have failing tests and conflicts to resolve

**Comment by eddyb** [comment]
@Dylan-DPC I've been doing that locally and haven't pushed, sorry (see also https://github.com/rust-lang/rust/pull/57967#issuecomment-476835081).

**Comment by eddyb** [comment]
@michaelwoerister @alexcrichton I've updated the PR description with some information about the new `-Z symbol-mangling` flag, and all the tests should be mangling-agnostic now.
(I've also tested with a randomized default mangling, and everything seemed fine)

So this is ready for review, and if there isn't any opposition to it, landing!

**Comment by michaelwoerister** [comment]
It looks like we don't have up-to-date performance numbers. @eddyb, could you do a perf.rlo run with the new scheme enabled?

**Comment by alexcrichton** [comment]
This all looks great to me! I'm happy whenever @michaelwoerister is

**Comment by michaelwoerister** [comment]
@rust-timer build 8c0711cf4f7f5d41798169eb18b77a5cfe138b93

**Comment by eddyb** [comment]
Heh, Travis is failing because `session::config::tests::test_debugging_options_tracking_hash` changes the default to `V0` and expects it to change (but the default is temporarily set to `V0`).
(i.e. there's nothing to do here, it's caused by the temporary commit for perf, and if we switch the default for real, we need to update the test too, to change from default to `Legacy`)

**Comment by michaelwoerister** [comment]
So, I see quite a few noticeable regressions. Up to ~6% for some of the `clean incremental` cases (`tokio-webpush-simple-debug`, `ripgrep-debug`, and `sentry-cli-debug`). It would be interesting to profile this, i.e. whether caching doesn't work properly or there is just more data to load with the longer symbol names. For the `clean incremental` case, we should not need to actually build any symbol names (iff we still cache them).

But other cases are affected too: `clean` and `baseline incremental` builds, although it's not consistent which of the cases is affected more. `baseline incremental` certainly will have more data to hash than before. 

The regressions don't block merging this PR, in my opinion, since we still default to the legacy scheme. But performance should be looked into some more before we do the final switch.

cc @rust-lang/wg-compiler-performance

File: src/librustc_codegen_utils/symbol_names/v0.rs
Comment (michaelwoerister) [inline src/librustc_codegen_utils/symbol_names/v0.rs]: Maybe add a comment of what this is about, i.e. that `_` represents `0` and `0_` represents `1`.

**Review by michaelwoerister — APPROVED** [review]
OK, this looks good. I think we should have more test cases in `src/test/ui/symbol-names`. Generics, closures, statics, etc. The reference implementation has [some examples](https://github.com/michaelwoerister/std-mangle-rs/blob/master/src/demangling_test_data.txt) (although I just saw that some of them are commented out `:/`)

Also, do we have tests somewhere that make sure that `rustc-demangle` produces the desired results for  the mangled symbols the compiler generates? Would be good to have some, so the two implementations don't go out of sync.

Other than that, I think we can merge this.

File: src/librustc/query/mod.rs
Comment (eddyb) [reply src/librustc/query/mod.rs]: I guess if we're only going to be doing versioning with this flag, and not different style, this would be fine.

**Comment by eddyb** [comment]
> Also, do we have tests somewhere that make sure that `rustc-demangle` produces the desired results for the mangled symbols the compiler generates? Would be good to have some, so the two implementations don't go out of sync.

I had that in the harness, but ofc that doesn't help us long-term. We can make the symbol name tests also print the `rustc-demangle` output, at least.

**Comment by eddyb** [comment]
@michaelwoerister I addressed both comments, I guess this is now only blocked on tests?
Maybe we should figure out what to do about https://github.com/rust-lang/rfcs/pull/2603#discussion_r279670424.

**Comment by michaelwoerister** [comment]
> We can make the symbol name tests also print the rustc-demangle output, at least.

Yes, that seems like a good solution.

**Comment by eddyb** [comment]
@michaelwoerister I just rebased this on #60525, to take advantage of https://github.com/rust-lang/rfcs/pull/2603#issuecomment-489232543, and to remove the weird hashing of the `DefPathData` variant.

**Comment by estebank** [comment]
CC #7539

**Comment by michaelwoerister** [comment]
@eddyb, do still plan to add `rustc-demangle` cross-checking to the `ui/symbol-names` tests?

**Comment by eddyb** [comment]
@michaelwoerister I've implemented the demangling testing and added another (weirder) testcase.
So I think this is ready to merge! (hopefully for real now)

**Comment by eddyb** [comment]
Oh, I guess this means we can't test paths from `core`/`std` because they will differ in hashes.

**Comment by nnethercote** [comment]
Valgrind currently is able to demangle Rust symbols. The code to do so is [here](https://sourceware.org/git/?p=valgrind.git;a=blob;f=coregrind/m_demangle/rust-demangle.c). Will that code need updating once this lands?

**Comment by eddyb** [comment]
@nnethercote That's a better question for the tracking issue, IMO.
This PR, as noted in the description, doesn't change the default mangling, precisely because tools are not updated (I plan to upstream [this code](https://gist.github.com/eddyb/c41a69378750a433767cf53fe2316768) to binutils/gdb, lldb, perf and valgrind).

**Comment by eddyb** [comment]
Oh, that's unfortunate, the bug that `std::ffi::VaList` leaks into `extern fn(X, ...)` types (see  https://github.com/rust-lang/rust/issues/44930#issuecomment-497163539) means I can't test `...` right now, I'll leave it out and then the test should be deterministic.

File: src/librustc_codegen_llvm/Cargo.toml
Comment (michaelwoerister) [inline src/librustc_codegen_llvm/Cargo.toml]: This should be `0.1.15` too, right?

File: src/librustc_codegen_ssa/Cargo.toml
Comment (michaelwoerister) [inline src/librustc_codegen_ssa/Cargo.toml]: This too?

File: src/tools/tidy/src/extdeps.rs
Comment (michaelwoerister) [inline src/tools/tidy/src/extdeps.rs]: This should be remove in the final version.

**Comment by michaelwoerister** [comment]
Alright, r=me with the nits addressed. Thanks, @eddyb!

File: src/tools/tidy/src/extdeps.rs
Comment (eddyb) [reply src/tools/tidy/src/extdeps.rs]: It should be, but maybe that change was lost in a rebase.

File: src/librustc_codegen_llvm/Cargo.toml
Comment (eddyb) [reply src/librustc_codegen_llvm/Cargo.toml]: Ah, oops, forgot to do that in one of the dozen rebases last night :D

**Comment by Centril** [comment]
@bors p=30

Putting above r0llup.

**Comment by michaelwoerister** [comment]
I don't know what's happening here. Doesn't reproduce locally for me.

**Comment by michaelwoerister** [comment]
OK, now I can reproduce locally but I don't really know what's going on. These are the files produced by the test:


### stable_symbol_names1_cross.nm
```
0000000000000010 T __ZN20stable_symbol_names113mono_function17ha52a8d588cef0d56E
0000000000000020 T __ZN20stable_symbol_names122mono_function_lifetime17h395d4c923e6c9632E
```

It looks like the `sed` script isn't working anymore?

### stable_symbol_names2_cross.nm
```
                 U __ZN20stable_symbol_names113mono_function17ha52a8d588cef0d56E
                 U __ZN20stable_symbol_names122mono_function_lifetime17h395d4c923e6c9632E
```

**Comment by michaelwoerister** [comment]
Ah, it's probably that macOS's `sed` doesn't like the new regex:
https://github.com/rust-lang/rust/pull/57967/files#diff-98aeca16a56f53e389304e8772f7ce2aR14

**Comment by eddyb** [comment]
The regex should be fixed now - @michaelwoerister found out that on macOS, non-`-E` `sed` understands `\(` and `\)` (at least), but not `\|`, which differs from GNU `sed` and even macOS non-`-E` `grep`!

@bors r=michaelwoerister

**Pull request merged**
